datatypes/src/operations/image/colorizer.rs

```diff
@@ -0,0 +1,532 @@
+
+                    let prev_color = breakpoint_prev.1;
+                    let next_color = breakpoint_next.1;
+
+                    let fraction = match self {
+                        Self::LinearGradient { .. } => {
+                            (value - prev_value) / (next_value - prev_value)
+                        }
+                        Self::LogarithmicGradient { .. } => {
+                            let nominator = f64::log10(value) - f64::log10(prev_value);
+                            let denominator = f64::log10(next_value) - f64::log10(prev_value);
+                            nominator / denominator
+                        }
+                        _ => unreachable!(), // cf. first match in function
+                    };
+
+                    prev_color.factor_add(next_color, fraction)
+                }
+            })
+            .collect();
+
+        debug_assert_eq!(color_table.len(), number_of_colors);
+
+        color_table
+    }
+}
+
+/// A ColorMapper is a function for mapping raster values to colors
+pub enum ColorMapper<'c> {
+    ColorTable {
+        color_table: Vec<RgbaColor>,
+        min_value: f64,
+        max_value: f64,
+        no_data_color: RgbaColor,
+        default_color: RgbaColor,
+    },
+    ColorMap {
+        color_map: &'c Palette,
+        no_data_color: RgbaColor,
+    },
+    Rgba,
+}
+
+// TODO: use Fn-trait once it is stable
+impl<'c> ColorMapper<'c> {
+    /// Map a raster value to a color from the colorizer
+    pub fn call<T>(&self, value: T) -> RgbaColor
+    where
+        T: Into<f64> + RgbaTransmutable,
+    {
+        match self {
+            ColorMapper::ColorTable {
+                color_table,
+                min_value,
+                max_value,
+                no_data_color,
+                default_color,
+            } => {
+                let value = value.into();
+                if f64::is_nan(value) {
+                    *no_data_color
+                } else if value < *min_value || value > *max_value {
+                    *default_color
+                } else {
+                    let color_table_factor = (color_table.len() - 1) as f64;
+                    let table_entry = f64::round(
+                        color_table_factor * ((value - *min_value) / (*max_value - *min_value)),
+                    ) as usize;
+                    *color_table.get(table_entry).unwrap_or(default_color)
+                }
+            }
+            ColorMapper::ColorMap {
+                color_map,
+                no_data_color,
+            } => {
+                let value = value.into();
+                if value.is_nan() {
+                    *no_data_color
+                } else {
+                    let value = unsafe { NotNan::unchecked_new(value) }; // checked by `if` above
+                    *color_map.get(&value).unwrap_or(no_data_color)
+                }
+            }
+            ColorMapper::Rgba => value.transmute_to_rgba(),
```

Comment:
since you transmute the original value directly, you should remove the transmute_to_f64 function from the trait.

Reply:
68fcf96